cpp/src/arrow/compute/kernels/compare.cc

```diff
@@ -17,250 +17,296 @@
-      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::LESS>(ctx);
+      return std::make_shared<CompareKernel<ArrowType, CompareOperator::LESS>>();
+
     case CompareOperator::LESS_EQUAL:
-      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::LESS_EQUAL>(ctx);
+      return std::make_shared<CompareKernel<ArrowType, CompareOperator::LESS_EQUAL>>();
   }
 
   return nullptr;
 }
 
-std::shared_ptr<CompareFunction> MakeCompareFunction(FunctionContext* ctx,
-                                                     const DataType& type,
-                                                     struct CompareOptions options) {
-  switch (type.id()) {
-    case UInt8Type::type_id:
-      return MakeCompareFunctionType<UInt8Type>(ctx, options);
-    case Int8Type::type_id:
-      return MakeCompareFunctionType<Int8Type>(ctx, options);
-    case UInt16Type::type_id:
-      return MakeCompareFunctionType<UInt16Type>(ctx, options);
-    case Int16Type::type_id:
-      return MakeCompareFunctionType<Int16Type>(ctx, options);
-    case UInt32Type::type_id:
-      return MakeCompareFunctionType<UInt32Type>(ctx, options);
-    case Int32Type::type_id:
-      return MakeCompareFunctionType<Int32Type>(ctx, options);
-    case UInt64Type::type_id:
-      return MakeCompareFunctionType<UInt64Type>(ctx, options);
-    case Int64Type::type_id:
-      return MakeCompareFunctionType<Int64Type>(ctx, options);
-    case FloatType::type_id:
-      return MakeCompareFunctionType<FloatType>(ctx, options);
-    case DoubleType::type_id:
-      return MakeCompareFunctionType<DoubleType>(ctx, options);
-    case Date32Type::type_id:
-      return MakeCompareFunctionType<Date32Type>(ctx, options);
-    case Date64Type::type_id:
-      return MakeCompareFunctionType<Date64Type>(ctx, options);
-    case TimestampType::type_id:
-      return MakeCompareFunctionType<TimestampType>(ctx, options);
-    case Time32Type::type_id:
-      return MakeCompareFunctionType<Time32Type>(ctx, options);
-    case Time64Type::type_id:
-      return MakeCompareFunctionType<Time64Type>(ctx, options);
-    default:
-      return nullptr;
+struct UnpackType {
+  Status Visit(const NullType& unreachable) { return Status::OK(); }
+
+  Status Visit(const BooleanType& t) {
+    *out_ = UnpackOperator<BooleanType>(options_.op);
+    return Status::OK();
   }
+
+  template <typename Numeric>
+  enable_if_number<Numeric, Status> Visit(const Numeric& t) {
+    *out_ = UnpackOperator<Numeric>(options_.op);
+    return Status::OK();
+  }
+
+  template <typename Temporal>
+  enable_if_temporal<Temporal, Status> Visit(const Temporal& t) {
+    *out_ = UnpackOperator<Temporal>(options_.op);
+    return Status::OK();
+  }
+
+  template <typename StringLike>
+  enable_if_base_binary<StringLike, Status> Visit(const StringLike& t) {
+    *out_ = UnpackOperator<StringLike>(options_.op);
+    return Status::OK();
+  }
+
+  Status Visit(const DictionaryType& t) { return NotImplemented(t); }
```

Comment:
Why don't you simply define a fallback `Status Visit(const DataType& t)`?

Reply:
I raised this issue while refactoring the type_traits. 

It's essentially equivalent to switch with a default case. If you add a new type (or pattern), you'll miss the visitor you need to implement (or ignore!) because it'll silently enter the catch-all pattern.

Thus by being explicit, whenever you add a new pattern, you need to take a decision to ignore or implement the visitors.

Reply:
To confirm: that was my intention as well